scripts/build-metrics.py

```diff
@@ -0,0 +1,296 @@
+            )
+            results.append(parsed_size)
+
+        return results
+
+
+class NinjaLogAdapter(BenchmarkAdapter):
+    """
+    Adapter to extract compile and link times from a .ninja_log.
+    Will calculate aggregates for total, compile, link and wall time.
+    Suite metadata will be set based on binary ending to object, library or executable.
+
+    Only files in paths beginning with velox/ will be tracked to avoid dependencies.
+    """
+
+    ninja_log: Path
+
+    def __init__(
+        self,
+        command: List[str],
+        ninja_log: str,
+        build_type: str,
+        result_fields_override: Dict[str, Any] = {},
+        result_fields_append: Dict[str, Any] = {},
+    ) -> None:
+        self.ninja_log = Path(ninja_log)
+        if build_type not in ["debug", "release"]:
+            raise ValueError(f"Build type '{build_type}' is not valid!")
+        self.build_type = build_type
+        super().__init__(command, result_fields_override, result_fields_append)
+
+    def _transform_results(self) -> List[BenchmarkResult]:
+        results = []
+
+        batch_id = uuid.uuid4().hex
+        with open(self.ninja_log, "r") as file:
+            log_lines = [line.strip() for line in file]
+
+        if not log_lines[0].startswith("# ninja log v"):
+            raise ValueError("Malformed Ninja log found!")
+        else:
+            del log_lines[0]
+
+        ms2sec = lambda x: x / 1000
+        get_epoch = lambda l: int(l.split()[2])
+        totals = {
+            "link_time": 0,
+            "compile_time": 0,
+            "total_time": 0,
+            "wall_time": get_epoch(log_lines[-1]) - get_epoch(log_lines[0]),
+        }
+
+        for line in log_lines:
+            start, end, epoch, object_path, _ = line.split()
+            start = int(start)
+            end = int(end)
+            duration = ms2sec(end - start)
+
+            # Don't track dependency times (refine check potentially?)
+            if not object_path.startswith("velox"):
+                continue
+
+            _, ext = splitext(object_path)
+            if ext in [".so", ".a"] or not ext:
+                totals["link_time"] += duration
+                suite = "linking"
+            elif ext == ".o":
+                totals["compile_time"] += duration
+                suite = "compiling"
+            else:
+                print(f"Unkown file type found: {object_path}")
+                print("Skipping...")
+                continue
+
+            time_result = BenchmarkResult(
+                run_reason="merge",
+                batch_id=batch_id,
+                stats={
+                    "data": [duration],
+                    "unit": "s",
+                    "iterations": 1,
+                },
+                tags={
+                    "name": object_path,
+                    "suite": suite,
+                    "source": f"{self.build_type}_build_metrics_time",
+                },
+                info={},
```

Comment:
i wonder if we can capture other things like ' compiler' , compiler flags used etc ? (maybe in the next set of changes).

Reply:
We can add arbitrary metadata as tags. There might also be a build in way to capture compiler and such similar to the machine info that's already part of the results.